src/Symfony/Component/HttpClient/CachingHttpClient.php

```diff
@@ -11,62 +11,346 @@
+use Symfony\Contracts\HttpClient\ChunkInterface;
 use Symfony\Contracts\HttpClient\HttpClientInterface;
 use Symfony\Contracts\HttpClient\ResponseInterface;
 use Symfony\Contracts\HttpClient\ResponseStreamInterface;
 use Symfony\Contracts\Service\ResetInterface;
 
 /**
- * Adds caching on top of an HTTP client.
+ * Adds caching on top of an HTTP client (per RFC 9111).
  *
- * The implementation buffers responses in memory and doesn't stream directly from the network.
- * You can disable/enable this layer by setting option "no_cache" under "extra" to true/false.
- * By default, caching is enabled unless the "buffer" option is set to false.
+ * Known omissions / partially supported features per RFC 9111:
+ *   1. Range requests:
+ *     - All range requests ("partial content") are passed through and never cached.
+ *   2. stale-while-revalidate:
+ *     - There's no actual "background revalidation" for stale responses, they will
+ *       always be revalidated.
+ *   3. min-fresh, max-stale, only-if-cached:
+ *     - Request directives are not parsed; the client ignores them.
  *
- * @author Nicolas Grekas <p@tchwork.com>
+ * @see https://www.rfc-editor.org/rfc/rfc9111
  */
 class CachingHttpClient implements HttpClientInterface, ResetInterface
 {
+    use AsyncDecoratorTrait {
+        stream as asyncStream;
+        AsyncDecoratorTrait::withOptions insteadof HttpClientTrait;
+    }
     use HttpClientTrait;
 
-    private HttpCache $cache;
+    /**
+     * The status codes that are always cacheable.
+     */
+    private const CACHEABLE_STATUS_CODES = [200, 203, 204, 206, 300, 301, 404, 405, 410, 414, 501];
+    /**
+     * The status codes that are cacheable if the response carry explicit cache directives.
+     */
+    private const CONDITIONALLY_CACHEABLE_STATUS_CODES = [302, 303, 307, 308];
+    /**
+     * The HTTP methods that are always cacheable.
+     */
+    private const CACHEABLE_METHODS = ['GET', 'HEAD'];
+    /**
+     * The HTTP methods that will trigger a cache invalidation.
+     */
+    private const UNSAFE_METHODS = ['POST', 'PUT', 'DELETE', 'PATCH'];
+
+    private TagAwareAdapterInterface|HttpCache $cache;
     private array $defaultOptions = self::OPTIONS_DEFAULTS;
 
+    /**
+     * @param bool     $sharedCache Indicates whether this cache is shared or private. When true, responses
+     *                              may be skipped from caching in presence of certain headers
+     *                              (e.g. Authorization) unless explicitly marked as public.
+     * @param int|null $maxTtl      The maximum time-to-live (in seconds) for cached responses.
+     *                              If a server-provided TTL exceeds this value, it will be capped
+     *                              to this maximum.
+     */
     public function __construct(
         private HttpClientInterface $client,
-        StoreInterface $store,
+        TagAwareAdapterInterface|StoreInterface $store,
```

Comment:
I suggest rename the argument to `$cache` so that the name matches the new API rather than the old one

Reply:
Adressed in https://github.com/symfony/symfony/pull/59576/commits/bac8c68cfb863328a36ccd4f70cd2147d7fab26e